hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/ProtobufHelper.java

```diff
@@ -46,4 +52,67 @@ public static IOException getRemoteException(ServiceException se) {
     }
     return e instanceof IOException ? (IOException) e : new IOException(se);
   }
+
+
+  /**
+   * Map used to cache fixed strings to ByteStrings. Since there is no
+   * automatic expiration policy, only use this for strings from a fixed, small
+   * set.
+   * <p/>
+   * This map should not be accessed directly. Used the getFixedByteString
+   * methods instead.
+   */
+  private static ConcurrentHashMap<Object, ByteString> fixedByteStringCache =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Get the ByteString for frequently used fixed and small set strings.
+   * @param key string
+   * @return
+   */
+  public static ByteString getFixedByteString(Text key) {
```

Comment:
Is there any risk this will grow to a size it becomes a problem? If not, using a guava cache is overkill, if there is a risk then yes, we would need to

Reply:
This is for frequently converting the fixed strings, such as 'token kind' and 'token service' in client side. These numbers will be fixed. ex: HDFS_DELEGATION_TOKEN, HDFS_BLOCK_TOKEN, etc.
For the other bytestrings (such as datanode network location) guava cache is used in PBHelperClient with expiry.
So it should be fine.